File: certification_server/src/models/commsec/commsec.go
package commsec

import (
	"strings"
	"time"
)

type CommKeyStatus string      // 通信密钥状态
type SignatureAlgorithm string // 签名算法
type CipherSuite string        // 密码套件

const (
	CommKeyActive  CommKeyStatus = "active"
	CommKeyExpired CommKeyStatus = "expired"
	CommKeyRevoked CommKeyStatus = "revoked"
)

const (
	SignatureECDSAP256SHA256 SignatureAlgorithm = "ecdsa_p256_sha256"
	SignatureEd25519         SignatureAlgorithm = "ed25519"
	SignatureRSAPSSSHA256    SignatureAlgorithm = "rsa_pss_sha256"
)

const (
	CipherSuiteAES128GCM        CipherSuite = "aes_128_gcm"
	CipherSuiteAES256GCM        CipherSuite = "aes_256_gcm"
	CipherSuiteChaCha20Poly1305 CipherSuite = "chacha20_poly1305"
)

// ServiceKeyOwner 标识通信密钥属于哪一个实体或实例。
// 统一使用 entity 语义字段，避免 service/entity 双轨冗余。
type ServiceKeyOwner struct {
	EntityType   string
	EntityID     string
	EntityName   string
	InstanceID   string
	InstanceName string
}

func (o ServiceKeyOwner) EffectiveEntityID() string {
	return o.EntityID
}

func (o ServiceKeyOwner) EffectiveEntityName() string {
	if o.EntityName != "" {
		return o.EntityName
	}
	return o.EntityID
}

func (o ServiceKeyOwner) Normalized() ServiceKeyOwner {
	o.EntityType = strings.ToLower(strings.TrimSpace(o.EntityType))
	o.EntityID = strings.TrimSpace(o.EntityID)
	o.EntityName = strings.TrimSpace(o.EntityName)
	o.InstanceID = strings.TrimSpace(o.InstanceID)
	o.InstanceName = strings.TrimSpace(o.InstanceName)
	if o.EntityName == "" {
		o.EntityName = o.EntityID
	}
	return o
}

// ServicePublicKeyRecord 表示存储在全局数据库中的实体通信公钥记录。
type ServicePublicKeyRecord struct {
	KeyID        string
	Owner        ServiceKeyOwner
	PublicKeyPEM string
	Fingerprint  string

	Status CommKeyStatus

	CreatedAt   time.Time
	ActivatedAt time.Time
	ExpiresAt   time.Time
	RevokedAt   time.Time
}

// LocalPrivateKeyRef 描述服务本地持有的私钥引用信息。
// 私钥本体不应进入数据库，只在本地安全存储中持有。
type LocalPrivateKeyRef struct {
	KeyID string
	Owner ServiceKeyOwner

	PrivateKeyRef string
	LoadedAt      time.Time
}

// PublicKeyLookupResult 是公钥目录查询结果。
type PublicKeyLookupResult struct {
	Found     bool
	Key       ServicePublicKeyRecord
	MatchedBy string

	FailureReason string
	CheckedAt     time.Time
}

// PublicKeyLookupRequest 表示统一的公钥目录查询请求。
type PublicKeyLookupRequest struct {
	KeyID string

	EntityID      string
	Owner         *ServiceKeyOwner
	RequireActive bool
}

func (r PublicKeyLookupRequest) Normalized() PublicKeyLookupRequest {
	r.KeyID = strings.TrimSpace(r.KeyID)
	r.EntityID = strings.TrimSpace(r.EntityID)
	if r.Owner != nil {
		normalizedOwner := r.Owner.Normalized()
		r.Owner = &normalizedOwner
		if r.EntityID == "" {
			r.EntityID = normalizedOwner.EffectiveEntityID()
		}
	}
	return r
}

File: data_worker/src/models/commsec/commsec.py
from __future__ import annotations

from typing import Literal

from msgspec import Struct


CommKeyStatus = Literal["active", "expired", "revoked"]
SignatureAlgorithm = Literal["ecdsa_p256_sha256", "ed25519", "rsa_pss_sha256"]
CipherSuite = Literal["chacha20_poly1305", "aes_256_gcm", "aes_128_gcm"]


class ServiceKeyOwner(Struct, kw_only=True):
    """表示通信密钥所有者，统一使用 entity 维度。"""

    entity_type: str = ""
    entity_id: str = ""
    entity_name: str = ""
    instance_id: str = ""
    instance_name: str = ""

    @property
    def effective_entity_id(self) -> str:
        return self.entity_id

    @property
    def effective_entity_name(self) -> str:
        return self.entity_name or self.entity_id

    def normalized(self) -> "ServiceKeyOwner":
        entity_type = self.entity_type.strip().lower()
        entity_id = self.entity_id.strip()
        entity_name = self.entity_name.strip() or entity_id
        return ServiceKeyOwner(
            entity_type=entity_type,
            entity_id=entity_id,
            entity_name=entity_name,
            instance_id=self.instance_id.strip(),
            instance_name=self.instance_name.strip(),
        )


class ServicePublicKeyRecord(Struct, kw_only=True):
    """表示一个服务公钥目录记录。"""

    key_id: str
    owner: ServiceKeyOwner

    public_key_pem: str
    fingerprint: str

    status: CommKeyStatus

    created_at: float
    activated_at: float
    expires_at: float
    revoked_at: float


class LocalPrivateKeyRef(Struct, kw_only=True):
    """表示一个本地私钥引用。"""

    key_id: str
    owner: ServiceKeyOwner

    private_key_ref: str
    loaded_at: float


class PublicKeyLookupRequest(Struct, kw_only=True):
    """统一公钥目录查询请求。"""

    key_id: str = ""
    entity_id: str = ""
    owner: ServiceKeyOwner | None = None
    require_active: bool = False

    def normalized(self) -> "PublicKeyLookupRequest":
        key_id = self.key_id.strip()
        entity_id = self.entity_id.strip()
        owner = self.owner.normalized() if self.owner is not None else None
        if owner is not None and not entity_id:
            entity_id = owner.effective_entity_id
        return PublicKeyLookupRequest(
            key_id=key_id,
            entity_id=entity_id,
            owner=owner,
            require_active=self.require_active,
        )


class PublicKeyLookupResult(Struct, kw_only=True):
    """公钥目录查询结果。"""

    found: bool
    key: ServicePublicKeyRecord | None = None
    matched_by: str = ""
    failure_reason: str = ""
    checked_at: float = 0.0
